### PR TITLE
Add custom time format implementations

### PR DESCRIPTION
**Pull request by straight-shoota**
The format patterns from Time::Format allow to define simple date formats but they are quite limited in flexibility. Many standardized formats include slight variations which must be taken into account for parsing such formats in a standards-compliant way.

This PR adds custom format implementations for some well-used time formats.
For a few it adds adds short-cut methods to `Time` class (like `Time.parse_iso8601`).

* `Time::Format::ISO_8601_DATETIME`: Parser for standard time format with many variations. It should support most of the standard format including week dates, ordinal dates and optional separators. Uses RFC 3339 as formatter.
* `Time::Format::RFC_3339`: More strict variation of ISO 8601, recommended for exchanging time data in computer systems.
* `Time::Format::RFC_2822`: Format compatible to RFC 822 and RFC 1123 used for example in HTTP protocols.
* `Time::Format::HTTP_DATE`: Proper implementation of `HTTP.parse_date` supporting time formats expressed as RFC 1123/RFC 2822, RFC 850 or asctime which are used by many HTTP protocols. Uses RFC_2822 as formatter.
* `Time::Format::YAML_DATE`: Refactored implementation of parser for YAML `datetime` type.

Surpasses #5084, #4729 

/cc @asterite @RX14

### CODE REVIEW

File: src/time/format/formatter.cr [hunk 1]
```diff
@@ -186,42 +222,69 @@ struct Time::Format
...
+    end
+
+    private def time_zone_gmt_or_rfc2822(**options)
+      case time.kind
+      when Time::Kind::Utc
+        time_zone_gmt
+      when Time::Kind::Local
+        time_zone_rfc2822
+      else
+        raise "invalid timezone"
```
Comment (Sija) [inline src/time/format/formatter.cr]: Capitalize exception message, please.

**Comment by straight-shoota** [comment]
I've rebased this PR and refactored a bit. It would be great to get some feedback.

**Comment by straight-shoota** [comment]
I've rebased this PR to incorporate the `Time::Location` API added in #5324.

Requires #5786

File: src/http/common.cr [hunk 2]
```diff
@@ -222,25 +220,38 @@ module HTTP
...
-    # TODO: GMT should come from the Time classes instead
-    time.to_utc.to_s("%a, %d %b %Y %H:%M:%S GMT")
+  #
+  # Uses `Time::Format::HTTP_DATE` as formatter.
+  def self.format_time(time : Time) : String
+    Time::Format::HTTP_DATE.format(time)
+  end
+
+  # DEPRECATED: Use `HTTP.format_time` instead.
+  def self.rfc1123_time(time : Time) : String
```
Comment (bcardiff) [inline src/http/common.cr]: Why this is left and marked as deprecated and rfc1123_date dropped completely? I think is either both or none.

File: src/json/to_json.cr [hunk 3]
```diff
@@ -124,7 +124,7 @@ end
 
 struct Time
   def to_json(json : JSON::Builder)
-    json.string(Time::Format::ISO_8601_DATE_TIME.format(self))
+    json.string(Time::Format::RFC_3339.format(self))
```
Comment (bcardiff) [inline src/json/to_json.cr]: Why this change if ISO_8601 is still defined?

File: spec/std/http/http_spec.cr [hunk 4]
```diff
@@ -37,15 +37,15 @@ describe HTTP do
     parsed_time.to_utc.to_s.should eq("2011-09-10 02:36:00 UTC")
   end
 
-  describe "generates RFC 1123" do
+  describe "generates HTTP date" do
     it "without time zone" do
       time = Time.utc(1994, 11, 6, 8, 49, 37, nanosecond: 0)
-      HTTP.rfc1123_date(time).should eq("Sun, 06 Nov 1994 08:49:37 GMT")
+      HTTP.format_time(time).should eq("Sun, 6 Nov 1994 08:49:37 GMT")
```
[the remark this answers is not in the document]
Comment (straight-shoota) [reply spec/std/http/http_spec.cr]: The name was changed because the methods doesn't implement one exact standard format but generally format and parse time instances for use in a HTTP protocol context (see previous comments in https://github.com/crystal-lang/crystal/pull/4729#issuecomment-317831282).

I'll look into the preferred digit format.

File: src/http/common.cr (hunk 2 above)
Comment (straight-shoota) [reply src/http/common.cr]: Yeah, it needs to be removed.

File: src/json/to_json.cr (hunk 3 above)
Comment (straight-shoota) [reply src/json/to_json.cr]: The difference is mostly negligible, but RFC 3339 is generally the better (stricter) format for this purpose. `ISO_8601_DATE_TIME` by default formats a offset as `HHmm` while RFC 3339 requires a colon: `HH:mm`

The JSON specification doesn't include a Date data type, so it's up to the implementation. JavaScript's `Date.toJSON()` actually is RFC 3339 compliant (`YYYY-MM-DDTHH:mm:ss.sssZ`).

**Comment by straight-shoota** [comment]
Thanks for taking a look at this!

1. Done.
2. Exactly. #5382 was supposed to be a quick bug fix, but it's been stalled for so long it would probably be better to just forget it and merge this one directly.
3. I don't think I had any further plans. Maybe I forgot... it's been a long time.

File: spec/std/json/serialization_spec.cr [hunk 5]
```diff
@@ -169,7 +169,7 @@ describe "JSON serialization" do
     end
 
     it "deserializes Time" do
-      Time.from_json(%("2016-11-16T09:55:48-0300")).to_utc.should eq(Time.utc(2016, 11, 16, 12, 55, 48))
+      Time.from_json(%("2016-11-16T09:55:48-03:00")).to_utc.should eq(Time.utc(2016, 11, 16, 12, 55, 48))
```
Comment (RX14) [inline spec/std/json/serialization_spec.cr]: We shouldn't break this

File: src/json/to_json.cr (hunk 3 above)
Comment (bcardiff) [reply src/json/to_json.cr]: The references I found said ISO8601 is the format used in Javascript. I think we should stick to 8601 here. Actually been less stricter is better IMO since an API built on top of this would be less pedantic to the user.

* https://www.w3schools.com/jsref/jsref_tojson.asp
* https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString

File: spec/std/json/serialization_spec.cr (hunk 5 above)
Comment (bcardiff) [reply spec/std/json/serialization_spec.cr]: According to https://github.com/crystal-lang/crystal/pull/5123#discussion_r193401753
I agree. The JSON format should be ISO8601 by default.

File: src/json/to_json.cr (hunk 3 above)
Comment (straight-shoota) [reply src/json/to_json.cr]: True, this is a bit confusing because RFC 3339 actually *is* ISO 8601. But RFC 3339 is better suited as machine-parsed exchange format. That's why I'm pretty convinced we should use RFC 3339 as output format for `#to_json`.

File: spec/std/json/serialization_spec.cr (hunk 5 above)
Comment (straight-shoota) [reply spec/std/json/serialization_spec.cr]: Output should be the RFC 3339 flavour of ISO 8601, the parser should accept more variants. That's probably the best solution.

File: src/json/to_json.cr (hunk 3 above)
Comment (straight-shoota) [reply src/json/to_json.cr]: I left `RFC_3339` here but it returns the exact same result as `ISO_8601_DATE_TIME`.

**Comment by bcardiff** [comment]
@straight-shoota 

1. Any further arguments on https://github.com/crystal-lang/crystal/pull/5123#discussion_r193403907 ? Why emitting in RFC3339 is better since the default in JavaScript is ISO8601? Luckily in utc they both use `Z` suffix. Although the added `:` is still ISO8601 compatible, we should be fine. @RX14 do you agree?

2. I noticed that now nanoseconds will be rendered in `to_json` while before only seconds were used. This was also aligned with [go behaviour](https://github.com/golang/go/blob/8da180f6cac51f6cb8a316b962f797dbaf7ee96f/src/time/format.go#L78) that includes a separata formatter for just seconds and nano seconds. Could you change the formatter to emit only seconds as before? I don't have a strong opinion to add a 3339Nano formatter but the extra precision might be problematic in some system where just milliseconds are supported.

After 2 is addressed is a 👍 from me (and a big 🙇 for all the time re work)

**Comment by straight-shoota** [comment]
1. There is no difference. `ISO_8601_DATE_TIME` would also use a colon in the offset, and that's how it should be. JavaScript's `Date.toJSON()` serializes in UTC only, by the way.
2. Good catch! We need more specs for these things 😆 But: should the default format entirely omit sub-second precision or round to milliseconds? (that's what  `Date.toJSON()` does). Though it probably won't be much of a difference, applications that understand milliseconds should probably be smart enough to digest a few extra digits.
There is no specification for any of this anyway, so we might as well use the best prevision available. If that doesn't work for some use case, it's easy to employ a custom formatter with a different format.

**Comment by bcardiff** [comment]
1. Ok
2. I would prefer the default to be the option that will broadly work. So seconds without fraction. If needed later a 3339Nano formatter can be. If a spec to ensure sub-second are drop is added definitely will be a plus

If you prefer me to made the changes in 2 let me know.

**Comment by straight-shoota** [comment]
Done. To accomplish this, I added a `fraction_digits` parameter to `RFC_3339.format` which I think would be generally useful and could be expanded. But that's probably something for a subsequent PR because this needs proper thought.

File: src/time/format.cr [hunk 6]
```diff
@@ -27,14 +30,15 @@
...
-# * **%N**: second fraction, zero padded. (Same as `%9N` but may consume more than 9 digits while parsing)
+# * **%L**: milliseconds, zero padded (000, 001, ..., 999)
 # * **%m**: month number, zero padded (01, 02, ..., 12)
 # * **%_m**: month number, blank padded (" 1", " 2", ..., "12")
 # * **%-m**: month number (1, 2, ..., 12)
 # * **%M**: minute, zero padded (00, 01, 02, ..., 59)
+# * **%3N**, **%L**: milliseconds, zero padded (000, 001, ..., 999)
+# * **%6N**: microseconds, zero padded (000000, 000001, ..., 999999)
+# * **%9N**: nanoseconds, zero padded (000000000, 000000001, ..., 999999999)
+# * **%N**: second fraction, zero padded. (Same as `%9N` but may consume more than 9 digits while parsing)
```
Comment (RX14) [inline src/time/format.cr]: Why this doc change?

File: src/time/format/formatter.cr [hunk 7]
```diff
@@ -23,6 +23,10 @@ struct Time::Format
       io << time.year / 100
     end
 
+    def full_or_short_year
+      year
```
Comment (RX14) [inline src/time/format/formatter.cr]: Why an alias?

File: src/time/format/formatter.cr [hunk 8]
```diff
@@ -79,6 +83,12 @@ struct Time::Format
       io << get_short_day_name.upcase
     end
 
+    def short_day_name_with_comma?
```
Comment (RX14) [inline src/time/format/formatter.cr]: This is a duplicate definition.

File: src/time/format.cr (hunk 6 above)
Comment (straight-shoota) [reply src/time/format.cr]: It helps discoverability if the directives are lexically ordered. It's unrelated, though. If you like, I can make a separate PR for this.

File: src/time/format/formatter.cr (hunk 7 above)
Comment (straight-shoota) [reply src/time/format/formatter.cr]: The parser has a different implementation.

File: src/time/format/formatter.cr (hunk 8 above)
Comment (straight-shoota) [reply src/time/format/formatter.cr]: Where's the duplicate?

**Comment by straight-shoota** [comment]
@RX14 You mean more like #5084? That design has been abandoned for good, I hope. This is the best I can think of and I don't see any problems with monkey patching a few specific methods nor any benefits in extracting them to modules.
We *could* put the specialized methods directly into the `formatter.cr`/`parser.cr` types to avoid monkey patching, but I think having these implementations in separate files is better code organization.

File: src/time/format/formatter.cr (hunk 8 above)
Comment (RX14) [reply src/time/format/formatter.cr]: Sorry I meant https://github.com/crystal-lang/crystal/pull/5123/files#diff-61fcbccde80bde3628b08077c5d0e22cR171 is overridden by https://github.com/crystal-lang/crystal/pull/5123/files#diff-819208e2dbfe2071e14f02023b5148d0R66.

File: src/time/format.cr [hunk 9]
```diff
@@ -27,14 +27,15 @@
...
-# * **%3N**, **%L**: milliseconds, zero padded (000, 001, ..., 999)
-# * **%6N**: microseconds, zero padded (000000, 000001, ..., 999999)
-# * **%9N**: nanoseconds, zero padded (000000000, 000000001, ..., 999999999)
-# * **%N**: second fraction, zero padded. (Same as `%9N` but may consume more than 9 digits while parsing)
+# * **%L**: milliseconds, zero padded (000, 001, ..., 999)
 # * **%m**: month number, zero padded (01, 02, ..., 12)
 # * **%_m**: month number, blank padded (" 1", " 2", ..., "12")
 # * **%-m**: month number (1, 2, ..., 12)
 # * **%M**: minute, zero padded (00, 01, 02, ..., 59)
+# * **%3N**, **%L**: milliseconds, zero padded (000, 001, ..., 999)
```
Comment (RX14) [inline src/time/format.cr]: duplicate definition of `L`

**Comment by straight-shoota** [comment]
They provide additional methods that are used by several specific formats. So there is added functionality for the core formatter/parser feature. Refactoring is actually in 0c242b0b.

**Comment by RX14** [comment]
Needs a rebase though

**Comment by bcardiff** [comment]
@straight-shoota I rebase the changes already.
